apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/RowAsyncEnumerator.cs

```diff
@@ -0,0 +1,298 @@
+
+                        // TODO: Capture response.LastScannedRowKey here for smart retries
+                        yield return _currentCell.Row;
+                    }
+                }
+            }
+        }
+
+        public bool IsRowInProgress => _rowMergeState != NewRow.Instance;
+
+        private void AddCompletedCellToRow()
+        {
+            var cell = new Cell
+            {
+                Value = _currentCell.Value,
+                TimestampMicros = _currentCell.Timestamp,
+                Labels = { _currentCell.Labels }
+            };
+            Assert(_currentCell.Column != null, "NewCell has no qualifier");
+            _currentCell.Column.Cells.Add(cell);
+        }
+
+        private void Assert(bool condition, string message)
+        {
+            if (!condition)
+            {
+                Reset();
+
+                // TODO: Is there something better we should throw here?
+                if (message != null)
+                {
+                    throw new InvalidOperationException(message);
+                }
+                else
+                {
+                    throw new InvalidOperationException();
+                }
+            }
+        }
+
+        private void Reset() => _currentCell = new CellInfo();
+
+        private struct CellInfo
+        {
+            public Column Column { get; set; }
+            public Family Family { get; set; }
+            public IEnumerable<string> Labels { get; set; }
+            public Row Row { get; set; }
+            public long Timestamp { get; set; }
+            public ByteString Value { get; set; }
+            public int ValueSizeExpected { get; set; }
+            public int ValueSizeRemaining { get; set; }
+        }
+
+        private abstract class RowMergeState
+        {
+            public abstract RowMergeState HandleChunk(RowAsyncEnumerator owner, CellChunk chunk);
+
+            public virtual RowMergeState CommitRow(RowAsyncEnumerator owner)
+            {
+                owner.Assert(false, null);
+                return this;
+            }
+        }
+
+        private sealed class NewRow : RowMergeState
+        {
+            public static readonly NewRow Instance = new NewRow();
+
+            private NewRow() { }
+
+            public override RowMergeState HandleChunk(RowAsyncEnumerator owner, CellChunk chunk)
+            {
+                owner.Assert(chunk.RowKey != null && !chunk.RowKey.IsEmpty, "NewRow must have a rowKey");
+                owner.Assert(chunk.FamilyName != null, "NewRow must have a family");
+                owner.Assert(chunk.Qualifier != null, "NewRow must have a qualifier");
+                owner.Assert(
+                    owner._currentCell.Row == null || BigtableByteString.Compare(owner._currentCell.Row.Key, chunk.RowKey) < 0,
+                    "NewRow key must be greater than the previous row's");
+
+                owner._currentCell.Row = new Row { Key = chunk.RowKey };
```

Comment:
Possibly add a warning against refactoring this to extract a local for `owner._currentCell`?

Reply:
Good call. Done.